File: implementations/csharp/src/Skight.AgentPlatform/WebTools.cs
using System;
using System.Net.Http;
using System.Text.Json;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace Skight.AgentPlatform
{
    public static class WebTools
    {
        private static readonly HttpClient HttpClient = new HttpClient();

        public static async Task<string> FetchUrlContentAsync(string argsJson)
        {
            try
            {
                using var doc = JsonDocument.Parse(argsJson);
                var root = doc.RootElement;
                if (!root.TryGetProperty("url", out var urlProp))
                {
                    return "Error: Missing 'url' argument.";
                }

                var url = urlProp.GetString()!;
                var response = await HttpClient.GetAsync(url);
                if (!response.IsSuccessStatusCode)
                {
                    return $"Error fetching URL ({response.StatusCode}): {response.ReasonPhrase}";
                }

                var html = await response.Content.ReadAsStringAsync();
                var text = StripHtml(html);
                return ToolSecurity.TruncateOutput(50_000, text);
            }
            catch (Exception ex)
            {
                return $"Error executing web_fetch_content: {ex.Message}";
            }
        }

        private static string StripHtml(string html)
        {
            if (string.IsNullOrWhiteSpace(html)) return string.Empty;
            // Remove script and style elements
            var clean = Regex.Replace(html, @"<(script|style)[^>]*?>.*?</\1>", "", RegexOptions.IgnoreCase | RegexOptions.Singleline);
            // Remove HTML tags
            clean = Regex.Replace(clean, @"<[^>]+>", " ");
            // Normalize whitespace
            return Regex.Replace(clean, @"\s+", " ").Trim();
        }
    }
}

File: implementations/fsharp/src/Skight.AgentPlatform.FSharp/WebTools.fs
namespace Skight.AgentPlatform.FSharp

open System
open System.Net.Http
open System.Text.Json
open System.Text.RegularExpressions
open System.Threading.Tasks

module WebTools =

    let private httpClient = new HttpClient()

    let private stripHtml (html: string) =
        if String.IsNullOrWhiteSpace(html) then ""
        else
            let clean1 = Regex.Replace(html, @"<(script|style)[^>]*?>.*?</\1>", "", RegexOptions.IgnoreCase ||| RegexOptions.Singleline)
            let clean2 = Regex.Replace(clean1, @"<[^>]+>", " ")
            Regex.Replace(clean2, @"\s+", " ").Trim()

    let fetchUrlContentAsync (argsJson: string) : Task<string> =
        async {
            try
                let doc = JsonDocument.Parse argsJson
                use _ = doc
                let root = doc.RootElement
                match root.TryGetProperty("url") with
                | true, urlProp ->
                    let url = urlProp.GetString()
                    let! response = httpClient.GetAsync(url) |> Async.AwaitTask
                    if not response.IsSuccessStatusCode then
                        return sprintf "Error fetching URL (%O): %s" response.StatusCode response.ReasonPhrase
                    else
                        let! html = response.Content.ReadAsStringAsync() |> Async.AwaitTask
                        let text = stripHtml html
                        return ToolSecurity.truncateOutput 50_000 text
                | false, _ ->
                    return "Error: Missing 'url' argument."
            with ex ->
                return sprintf "Error executing web_fetch_content: %s" ex.Message
        } |> Async.StartAsTask
